modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/memoize.js

```diff
@@ -0,0 +1,48 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+export default function memoize(fn, resolver) {
+
+	/**
+	 * Using the Map to enable storing objects as keys
+	 */
+	const cache = new Map();
+
+	/**
+	 * Returns a memoized function.
+	 * Takes the passed function and adds cache handling to it,
+	 * applying the passed arguments to maintain expected functionality
+	 */
+	return (...args) => {
+
+		/**
+		 * Use the resolver function if it is provided, otherwise use the first argument
+		 */
+		const key = resolver ? resolver.apply(null, args) : args[0];
+
+		/**
+		 * Check if the first argument is cached, if yes, return it, otherwise store it
+		 */
+		if (cache.has(key)) {
+			return cache.get(key);
+		}
+		else {
+			const result = fn.apply(null, args);
```

Comment:
Can you walk me through why you used `apply` here?

Reply:
We need to invoke the provided `fn` function somehow and accept arguments, so I figured that `apply` is the best at doing that.

This is the example usage which probably wasn't clear from the implementation:
```js
function randomFn(n) {return n};

const memoizedFn = memorize(randomFn);

memoizedFn(5);
```